packages/swapper/src/swappers/ButterSwap/swapperApi/getTradeQuote.ts

```diff
@@ -0,0 +1,199 @@
+        message: `[getTradeQuote] ${routeResponse.message}`,
+        code: butterSwapErrorToTradeQuoteError(routeResponse.errno),
+      }),
+    )
+  }
+
+  const route = routeResponse.data[0]
+  if (!route) {
+    return Err(
+      makeSwapErrorRight({
+        message: '[getTradeQuote] No route found',
+        code: TradeQuoteError.NoRouteFound,
+      }),
+    )
+  }
+
+  // Call ButterSwap /swap API to get calldata and contract info
+  const buildTxResult = await getBuildTx(
+    route.hash,
+    slippage,
+    sendAddress, // from (source chain address)
+    receiveAddress, // receiver (destination chain address)
+  )
+
+  if (buildTxResult.isErr()) return Err(buildTxResult.unwrapErr())
+  const buildTxResponse = buildTxResult.unwrap()
+  if (!isBuildTxSuccess(buildTxResponse)) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getTradeQuote] /swap failed: ${buildTxResponse.message}`,
+        code: TradeQuoteError.QueryFailed,
+      }),
+    )
+  }
+  const buildTx = buildTxResponse.data[0]
+  if (!buildTx) {
+    return Err(
+      makeSwapErrorRight({
+        message: '[getTradeQuote] No buildTx data returned',
+        code: TradeQuoteError.QueryFailed,
+      }),
+    )
+  }
+
+  // Fee asset for network/protocol fees
+  const feeAsset = _deps.assetsById[feeAssetId]
+  if (!feeAsset) {
+    return Err(
+      makeSwapErrorRight({
+        message: `[getTradeQuote] Fee asset not found for chainId ${sellAsset.chainId}`,
+        code: TradeQuoteError.UnsupportedChain,
+      }),
+    )
+  }
+
+  // Map gasFee.amount to networkFeeCryptoBaseUnit using fee asset precision
+  const networkFeeCryptoBaseUnit = toBaseUnit(bnOrZero(route.gasFee?.amount), feeAsset.precision)
+
+  // Calculate rate as lastHop.totalAmountOut / srcChain.totalAmountIn (in base units)
+  // For cross-chain swaps, use dstChain.totalAmountOut (final min amount out)
+  // For same-chain swaps, use srcChain.totalAmountOut
+  // We do NOT use bridgeChain.totalAmountOut, as it is only an intermediary for cross-chain swaps
+  const inputAmount = bnOrZero(route.srcChain.totalAmountIn)
+  const outputAmount = route.dstChain?.totalAmountOut ?? route.srcChain.totalAmountOut
+  const rate = inputAmount.gt(0) ? bnOrZero(outputAmount).div(inputAmount).toString() : '0'
+
+  const step = {
+    buyAmountBeforeFeesCryptoBaseUnit: toBaseUnit(outputAmount, buyAsset.precision),
+    buyAmountAfterFeesCryptoBaseUnit: toBaseUnit(outputAmount, buyAsset.precision),
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: amount,
+    feeData: {
+      networkFeeCryptoBaseUnit,
+      protocolFees: undefined,
+    },
+    rate,
+    source: SwapperName.ButterSwap,
+    buyAsset,
+    sellAsset,
+    accountNumber,
+    allowanceContract: route.contract ?? '0x0',
+    estimatedExecutionTimeMs: route.timeEstimated * 1000,
+    butterSwapTransactionMetadata: {
+      to: buildTx.to,
+      data: buildTx.data,
+      value: buildTx.value,
+      chainId: buildTx.chainId,
```

Comment:
ditto unused, rm from `butterSwapTransactionMetadata` types and here

Reply:
https://github.com/shapeshift/web/pull/9799/commits/44659376bba89d19a685dbea974e561fcde198cd